### PR TITLE
Enable Gradient Checkpointing for UNet2DModel

### PR DESCRIPTION
**Pull request by dg845**
# What does this PR do?

This PR enables gradient checkpointing for `UNet2DModel` by setting the `_supports_gradient_checkpointing` flag to `True`. Since `UNet2DConditionModel` has `_supports_gradient_checkpointing = True`, it seems like `UNet2DModel` should support gradient checkpointing as well, unless I'm missing something.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten
@sayakpaul

### CODE REVIEW

**Review by sayakpaul — COMMENTED** [review]
Do we not have also configure the gradient checkpointing blocks like how we do [here](https://github.com/huggingface/diffusers/blob/0d927c75427d12210de2ff3eb6fff1c4ebd2e130/src/diffusers/models/unets/unet_2d_condition.py#L760)?

**Comment by dg845** [comment]
> Do we not have also configure the gradient checkpointing blocks like how we do [here](https://github.com/huggingface/diffusers/blob/0d927c75427d12210de2ff3eb6fff1c4ebd2e130/src/diffusers/models/unets/unet_2d_condition.py#L760)?

You're right, I missed this 😅.

**Comment by dg845** [comment]
The `UNetMidBlock2D`, `AttnDownBlock2D`, and `AttnUpBlock2D` blocks used in `UNet2DModel` currently do not have gradient checkpointing implemented, so I have added gradient checkpointing to each of those blocks.

As a note, in their current `forward` method, `AttnDownBlock2D` and `AttnUpBlock2D` use the `scale` keyword argument when calling `resnet`, e.g.:

https://github.com/huggingface/diffusers/blob/d4c7ab7bf1a00b8f416b3d20b77babac86f7fb44/src/diffusers/models/unets/unet_2d_blocks.py#L1045-L1046

So I have written the `create_custom_forward` function as

https://github.com/huggingface/diffusers/blob/e83785732373bd3d0f4151a28c6a4cc95e90db94/src/diffusers/models/unets/unet_2d_blocks.py#L1072-L1079

This has the potential to cause problems if `return_dict` is also supplied through `kwargs` (see https://docs.python.org/3/tutorial/controlflow.html#function-examples).

`CrossAttnDownBlock2D` handles this by omitting the `scale` argument when calling `custom_forward` entirely:

https://github.com/huggingface/diffusers/blob/d4c7ab7bf1a00b8f416b3d20b77babac86f7fb44/src/diffusers/models/unets/unet_2d_blocks.py#L1183-L1188

which seems wrong when `lora_scale` is not `ResnetBlock2D.forward`'s default `scale` value of `1.0` because the forward passes with and without gradient checkpointing are not equivalent.

Since `scale` is currently the third keyword argument to `ResnetBlock2D.forward`, we can probably supply it as a positional argument and use `CrossAttnDownBlock2D`'s `create_custom_forward` implementation. I'm not sure which approach is best.

**Comment by sayakpaul** [comment]
> which seems wrong when lora_scale is not ResnetBlock2D.forward's default scale value of 1.0 because the forward passes with and without gradient checkpointing are not equivalent.

I think this is still fine because `lora_scale` is never going to interfere during training. This is still reasonable to me because there is always a bit of subtle differences for training and inference in the forward of a model. WDYT?

File: tests/models/test_models_unet_2d.py [hunk 1]
```diff
@@ -105,6 +107,84 @@ def test_mid_block_attn_groups(self):
...
+        init_dict["attention_head_dim"] = 8
+
+        model_class_copy = copy.copy(self.model_class)
+
+        modules_with_gc_enabled = {}
+
+        # now monkey patch the following function:
+        #     def _set_gradient_checkpointing(self, module, value=False):
+        #         if hasattr(module, "gradient_checkpointing"):
+        #             module.gradient_checkpointing = value
```
Comment (sayakpaul) [inline tests/models/test_models_unet_2d.py:164]: Can be removed.

File: tests/models/test_models_unet_2d.py [hunk 2]
```diff
@@ -105,6 +107,84 @@ def test_mid_block_attn_groups(self):
...
+
+        model = model_class_copy(**init_dict)
+        model.enable_gradient_checkpointing()
+
+        EXPECTED_SET = {
+            "AttnUpBlock2D",
+            "AttnDownBlock2D",
+            "UNetMidBlock2D",
+            "UpBlock2D",
+            # "Transformer2DModel",
```
Comment (sayakpaul) [inline tests/models/test_models_unet_2d.py:181]: Can be removed.

File: src/diffusers/models/unets/unet_2d_blocks.py [hunk 3]
```diff
@@ -2249,9 +2298,31 @@ def forward(
...
+
+                ckpt_kwargs: Dict[str, Any] = {"use_reentrant": False} if is_torch_version(">=", "1.11.0") else {}
+                hidden_states = torch.utils.checkpoint.checkpoint(
+                    create_custom_forward(resnet),
+                    hidden_states,
+                    temb,
+                    scale=scale,
+                    **ckpt_kwargs,
+                )
+                cross_attention_kwargs = {"scale": scale}
```
Comment (sayakpaul) [inline src/diffusers/models/unets/unet_2d_blocks.py:2320]: Can initialize it once and reuse.

**Review by sayakpaul — APPROVED** [review]
Just some nits, but looks very good. Nice test, too.

**Comment by dg845** [comment]
Regarding https://github.com/huggingface/diffusers/pull/6718#issuecomment-1912937574: I think in this case the best short term solution is to use the standard `create_custom_forward` implementation and supply the `lora_scale` as a positional argument when calling `resnet` using the `custom_forward` function. This allows the gradient checkpointing forward pass to be the same as the non-gradient checkpointing forward pass during training (which is also the same as the forward pass during inference). This does introduce some dependency on the order of the positional arguments in `ResnetBlock2D.forward`, but I think it's probably fine since the `ResnetBlock2D` API is likely to remain stable over time.

In the long term, at least in `src/diffusers/models/unets/unet_2d_blocks.py`, I think it might make sense to revisit the `create_custom_forward` implementation. A quick search through that file indicates that `create_custom_forward` is never called with the `return_dict` keyword argument. So perhaps a more general implementation like

```python
def create_custom_forward(module):
    def custom_forward(*inputs, **kwargs):
        return module(*inputs, **kwargs)

    return custom_forward
```

could be used, and `return_dict` could be supplied through `kwargs` if necessary: for example, if we're using [`torch.utils.checkpoint.checkpoint`](https://pytorch.org/docs/stable/checkpoint.html#torch.utils.checkpoint.checkpoint), something like 

```python
ckpt_kwargs: Dict[str, Any] = {"use_reentrant": False} if is_torch_version(">=", "1.11.0") else {}
hidden_states = torch.utils.checkpoint.checkpoint(
    create_custom_forward(resnet),
    hidden_states,
    temb,
    scale=lora_scale,
    return_dict=True,
    **ckpt_kwargs,
)
```

Note that [`torch.utils.checkpoint.checkpoint`](https://pytorch.org/docs/stable/checkpoint.html#torch.utils.checkpoint.checkpoint) supports both positional and keyword arguments, at least since v1.11.

**Comment by sayakpaul** [comment]
@dg845 I think I don't quite follow the concern fully. 

Could you maybe try to demonstrate the issue with a simpler example? 

> which seems wrong when lora_scale is not ResnetBlock2D.forward's default scale value of 1.0 because the forward passes with and without gradient checkpointing are not equivalent.

Would like to see when this case arises. From what I understand, gradient checkpointing is used _during_ training, and `lora_scale` is never supposed to be supplied during training. So, I don't quite understand how a discrepancy stems here. Maybe I am missing something. 

I would like to keep the legacy blocks as is until and unless absolutely necessary. This is why I am asking for a simpler example to understand the consequences.

**Comment by dg845** [comment]
> which seems wrong when lora_scale is not ResnetBlock2D.forward's default scale value of 1.0 because the forward passes with and without gradient checkpointing are not equivalent.

Sorry, I should have made it clear that the above follows from my belief that the `lora_scale` _should_ be supplied during training.

My understanding is that in the [original LoRA paper](https://arxiv.org/pdf/2106.09685.pdf) the LoRA scale parameter $`\alpha`$ is a hyperparameter during training:

lora_alpha_param

I think in practice $`\alpha`$ is typically held constant and the learning rate is tuned during training (following the highlighted section), but theoretically we could treat $`\alpha`$ and the learning rate as independent hyperparameters and tune them both.

Similarly, if we look at [`peft.tuners.lora.layer.Linear`](https://github.com/huggingface/peft/blob/bfc102c0c095dc9094cdd3523b729583bfad4688/src/peft/tuners/lora/layer.py#L191), the [`forward`](https://github.com/huggingface/peft/blob/bfc102c0c095dc9094cdd3523b729583bfad4688/src/peft/tuners/lora/layer.py#L301) method does not disable scaling the LoRA update $`\Delta W`$ during training:

https://github.com/huggingface/peft/blob/bfc102c0c095dc9094cdd3523b729583bfad4688/src/peft/tuners/lora/layer.py#L318-L320

unlike for something like [dropout](https://pytorch.org/docs/stable/generated/torch.nn.Dropout.html#torch.nn.Dropout) where the forward pass would be different depending on whether `torch.nn.Module.training` is set.

So in my view the discrepancy between the gradient checkpointing code and non-gradient checkpointing code in e.g. `CrossAttnDownBlock2D` when `lora_scale != 1.0` is a bug because

1. The gradient checkpointing forward pass differs from the non-gradient checkpointing forward pass during training.
2. For the same reason, the gradient checkpointing forward pass during training differs from the forward pass during inference, which will result in a train-test mismatch if we train using gradient checkpointing.

Practically speaking, we might not consider the train-test mismatch that arises to be that bad, since we may want to tune the scaling of the LoRA update during inference anyway (e.g. if we are performing inference with multiple LoRAs simultaneously).

**Comment by dg845** [comment]
That being said, perhaps it's better if I move the changes (especially to `CrossAttnDownBlock2D` and `CrossAttnUpBlock2D` in 8756be5) out of this PR, and this can be revisited in a separate issue/PR.

**Comment by sayakpaul** [comment]
`alpha` is not the same as the `scale` parameter in LoRA training in my understanding. `scale` is an inference-time parameter and shouldn't influence training whereas `alpha` could be tuned during training of LoRAs. For LoRA training, we rely on PEFT. If we want to expose `alpha` for training we can easily do so. Example:

https://github.com/huggingface/diffusers/blob/5b1b80a5b6633126994636ea289393b9e931834a/examples/dreambooth/train_dreambooth_lora_sdxl.py#L987

But this discussion is starting to deviate from the original topic of the PR a bit IMO. 

> The gradient checkpointing forward pass differs from the non-gradient checkpointing forward pass during training.
> For the same reason, the gradient checkpointing forward pass during training differs from the forward pass during inference, which will result in a train-test mismatch if we train using gradient checkpointing.

^ this I agree. And maybe this could be handled first in a separate PR and then we revisit this PR. Does that work?

**Comment by dg845** [comment]
> And maybe this could be handled first in a separate PR and then we revisit this PR. Does that work?

Sounds good :). To be more precise, would something like this sound good to you?

1. In this PR, gradient checkpointing is implemented for `UNet2DModel` and its associated blocks such as `AttnDownBlock2D`/`AttnUpBlock2D` in a way which is exactly parallel to the current gradient checkpointing implementation in `UNet2DConditionModel` and `CrossAttnDownBlock2D`/`CrossAttnUpBlock2D`.
2. The question of how the `scale` parameter should be handled for the legacy LoRA implementation will be revisited in a separate issue and/or PR.

**Comment by sayakpaul** [comment]
Yeah that is right.

**Comment by dg845** [comment]
I have updated the gradient checkpointing implementation in this PR to be exactly parallel to that of `UNet2DConditionModel` and opened a new issue regarding the LoRA scale parameter `scale` at #6756.

**Review by sayakpaul — APPROVED** [review]
Thanks a mile!

**Comment by sayakpaul** [comment]
@yiyixuxu could you check it as well?

File: src/diffusers/models/unets/unet_2d_blocks.py [hunk 4]
```diff
@@ -1042,10 +1067,32 @@ def forward(
...
+                        if return_dict is not None:
+                            return module(*inputs, return_dict=return_dict)
+                        else:
+                            return module(*inputs)
+
+                    return custom_forward
+
+                ckpt_kwargs: Dict[str, Any] = {"use_reentrant": False} if is_torch_version(">=", "1.11.0") else {}
+                cross_attention_kwargs.update({"scale": lora_scale})
+                hidden_states = torch.utils.checkpoint.checkpoint(
```
Comment (yiyixuxu) [inline src/diffusers/models/unets/unet_2d_blocks.py:1083]: I don't think `scale=lora_scale` argument is being passed down to `resnet()` when checkpointing? did I miss anything? 
cc @sayakpaul

**Review by yiyixuxu — COMMENTED** [review]
Thanks for working on this! 
I left a question:)

File: src/diffusers/models/unets/unet_2d_blocks.py [hunk 5]
```diff
@@ -1042,10 +1067,32 @@ def forward(
...
+            if self.training and self.gradient_checkpointing:
+
+                def create_custom_forward(module, return_dict=None):
+                    def custom_forward(*inputs):
+                        if return_dict is not None:
+                            return module(*inputs, return_dict=return_dict)
+                        else:
+                            return module(*inputs)
+
+                    return custom_forward
```
Comment (yiyixuxu) [inline src/diffusers/models/unets/unet_2d_blocks.py:1079]: would something like this work?
```suggestion
                def create_custom_forward(module, scale=None):
                    def custom_forward(*inputs):
                        if scale is not None:
                            return module(*inputs, scale=lora_scale)
                        else:
                            return module(*inputs)

                    return custom_forward
```

Comment (dg845) [reply src/diffusers/models/unets/unet_2d_blocks.py:1079]: My current understanding is that the `scale` argument should only be supplied at inference time, not both at training time and inference time as I originally thought. Since `custom_forward` is only used when gradient checkpointing during training, I'm don't think it should be called with a non-default `scale` argument.

For this PR, I would like to keep the `UNet2DModel`/`AttnDownBlock2D`/... gradient checkpointing logic as close to the current gradient checkpointing logic in `UNet2DConditionModel`/`CrossAttnDownBlock2D`/..., following https://github.com/huggingface/diffusers/pull/6718#issuecomment-1913574196 and https://github.com/huggingface/diffusers/pull/6718#issuecomment-1913846617, and implement general changes to the `lora_scale` handling logic in #6756 and any follow up PRs.

Comment (yiyixuxu) [reply src/diffusers/models/unets/unet_2d_blocks.py:1079]: ok :)

**Review by yiyixuxu — APPROVED** [review]

**Comment by dg845** [comment]
Created a new PR with the changes at #7201. Will close this PR.
